test/build.bats

```diff
@@ -67,60 +67,82 @@ load common
+    # the Bats tests forever:
+    #
+    #  SC1090  can't find sourced file
+    #  SC2154  variable referenced but not assigned
+    #
     scope standard
-    ( command -v shellcheck >/dev/null 2>&1 ) || skip "no shellcheck found"
-    # skip if minimum shellcheck not met unless Travis or LANL
+    # Only do this test in build directory; the reasoning is that we don't
+    # alter the shell scripts during install enough to re-test, and it means
+    # we only have to find everything in one path.
+    if [[ $CHTEST_INSTALLED ]]; then
+        skip 'only in build directory'
+    fi
+    # Are we on Travis or at LANL?
+    if [[ $TRAVIS || $(hostname --fqdn) = *'.lanl.gov' ]]; then
+        pedantic=yes
+    else
+        pedantic=
+    fi
+    # ShellCheck present?
+    if ( ! command -v shellcheck >/dev/null 2>&1 ); then
+        error='no shellcheck found'
+        if [[ $pedantic ]]; then
+             echo "$error"
+             false
+        else
+             skip "$error"
+        fi
+    fi
+    # ShellCheck minimum version?
     version=$(shellcheck --version | grep -E '^version:' | cut -d' ' -f2)
     major=${version%%.*}
     rest=${version#*.}
     minor=${rest%%.*}
-    echo "shellcheck: version '$version', major '$major', minor '$minor'"
-    if [[ $minor -lt 6 ]]; then
+    echo "shellcheck: version '${version}', major '${major}', minor '${minor}'"
+    minor_needed=6
+    if [[ $minor -lt $minor_needed ]]; then
         # no need to check major because minimum is 0
-        error="shellcheck $version older than 0.6.0"
-        if [[ $TRAVIS ]] || [[ $(hostname --fqdn) = *'.lanl.gov' ]]; then
+        error="shellcheck ${version} older than 0.${minor_needed}.0"
+        if [[ $pedantic ]]; then
             echo "$error"
             false
         else
             skip "$error"
         fi
     fi
-    # user executables
-    for i in "$ch_bin"/ch-*; do
+    # Shell scripts and libraries: appropriate extension or shebang.
+    # For awk program, see: https://unix.stackexchange.com/a/66099
+    while IFS= read -r i; do
         echo "shellcheck: ${i}"
-        [[ ! $(file "$i") = *'shell script'* ]] && continue
         shellcheck -e SC1090,SC2002,SC2154 "$i"
-    done
-    # libraries for user executables
-    for i in "$ch_libexec"/*.sh; do
-        echo "shellcheck: ${i}"
-        shellcheck -s sh -e SC1090,SC2002 "$i"
-    done
-    # BATS scripts
+    done < <( find "$ch_base" \
+                   \(    -name .git \
+                      -o -name build-aux \) -prune \
+                -o \( -name '*.sh' -print \) \
+                -o \( -name '*.bash' -print \) \
+                -o \( -type f -exec awk '/^#!\/bin\/(ba)?sh/ {print FILENAME}
+                                         {nextfile}' {} + \) )
+    # Bats scripts. Use sed to do two things:
+    #
+    # 1. Make parseable by ShellCheck by removing "@test '...'". This does
+    #    remove the test names, but line numbers are still valid.
```

Comment:
What is the effect of removing the test names?

Reply:
ShellCheck warnings will not be quite as clear. But the line numbers are correct, so you can still go look easily.